### PR TITLE
rename variable link to mediaUrl

### LINKED ISSUE #1581: Rename variable link to mediaUrl
**Is your feature request related to a problem? Please describe.**
A clear and concise description of what the problem is.

Recently we've added support for uploading media instead of accepting a link when creating an advertisement to minimise chances of error. 

https://github.com/PalisadoesFoundation/talawa-admin/pull/1248

As discussed [here](https://github.com/PalisadoesFoundation/talawa-admin/pull/1248#issuecomment-1869562680), the link parameter needs to be renamed to mediaUrl.

**Describe the solution you'd like**
Rename the variable link to mediaUrl

**Describe alternatives you've considered**

**Approach to be followed (optional)**
A clear and concise description of approach to be followed.

**Additional context**

**Potential internship candidates**
Please read this if you are planning to apply for a Palisadoes Foundation internship https://github.com/PalisadoesFoundation/talawa/issues/359

### PR DESCRIPTION
**Pull request by disha1202**
**What kind of change does this PR introduce?**

feature

**Issue Number:**

Fixes #1581 

**Did you add tests for your changes?**
No

**Snapshots/Videos:**

**If relevant, did you update the documentation?**

**Summary**

**Does this PR introduce a breaking change?**

**Other information**

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**

### CODE REVIEW

**Comment by palisadoes** [comment]
Is this suitable now that we are planning on not using a URL for advertising?

**Comment by disha1202** [comment]
> Is this suitable now that we are planning on not using a URL for advertising?

It was working fine with string type as well, but I have updated it to URL for code consistency and also verified the changes by creating advertisement. 

Please take a look.

**Comment by palisadoes** [comment]
1. The aim is to upload an image not use a URL. 
1. Why is the URL methodology being used?

**Comment by disha1202** [comment]
> * The aim is to upload an image not use a URL.
> * Why is the URL methodology being used?

Oh Okay, I might have misunderstood earlier. 
Yes this is suitable for uploading the image/video in advertisement instead of URL. 

And by URL methodology you mean the type [URL](https://github.com/PalisadoesFoundation/talawa-api/pull/1585/files#diff-efc1675187620595b0844197a25c35eac9b6df752f9182e5cffddee6f27a8489R8)?

**Comment by palisadoes** [comment]
@xoldyckk please take a look

**Comment by palisadoes** [comment]
> > * The aim is to upload an image not use a URL.
> > * Why is the URL methodology being used?
> 
> Oh Okay, I might have misunderstood earlier. 
> Yes this is suitable for uploading the image/video in advertisement instead of URL. 
> 
> And by URL methodology you mean the type [URL](https://github.com/PalisadoesFoundation/talawa-api/pull/1585/files#diff-efc1675187620595b0844197a25c35eac9b6df752f9182e5cffddee6f27a8489R8)?
> 

My concern is that it seems that you are using a variable named mediaURL to reference a file upload. Is going to be confusing to future developers.

Also, is the file stored using the standard methodology for all other uploads? For example, post images.

**Comment by palisadoes** [comment]
OK. We'll need to fix this later.

File: schema.graphql [hunk 1]
```diff
@@ -438,7 +438,7 @@ type Mutation {
   cancelMembershipRequest(membershipRequestId: ID!): MembershipRequest!
   checkIn(data: CheckInInput!): CheckIn!
   createAdmin(data: UserAndOrganizationInput!): User!
-  createAdvertisement(endDate: Date!, link: String!, name: String!, orgId: ID!, startDate: Date!, type: String!): Advertisement!
+  createAdvertisement(endDate: Date!, mediaUrl: URL, name: String!, orgId: ID!, startDate: Date!, type: String!): Advertisement!
```
Comment (xoldyckk) [inline schema.graphql:441]: Why is `mediaUrl` not required? Can an advertisement be made without it?

Also in mutations it's better to have them accept a single input object and define an input graphql type for it, here for example something like `input CreateAdvertisementInput`, having seperate fields littered around like this just makes the schema design inconsistent.

**Comment by disha1202** [comment]
Hi @xoldyckk 

I have made the suggested changes. 
Please take a look.

**Comment by disha1202** [comment]
Any update on this?

File: src/typeDefs/mutations.ts [hunk 2]
```diff
@@ -88,14 +88,7 @@ export const mutations = gql`
...
 
-    createAdvertisement(
-      orgId: ID!
-      name: String!
-      link: String!
-      type: String!
-      startDate: Date!
-      endDate: Date!
-    ): Advertisement!
+    createAdvertisement(data: createAdvertisementInput!): Advertisement!
```
Comment (xoldyckk) [inline src/typeDefs/mutations.ts:91]: `input: CreateAdvertisementInput!`

File: src/models/Advertisement.ts [hunk 3]
```diff
@@ -11,7 +11,7 @@ export interface InterfaceAdvertisement {
   _id: Types.ObjectId;
   orgId: string;
   name: string;
-  link: string;
+  mediaUrl: string | undefined | null;
```
Comment (xoldyckk) [inline src/models/Advertisement.ts:14]: `mediaUrl` is required in mongoose and graphql schema, why would you put the typescript type to be nullable?

**Review by xoldyckk — COMMENTED** [review]
`PascalCase` for inputs, types, enums in graphql schema

File: src/resolvers/Mutation/createAdvertisement.ts [hunk 4]
```diff
@@ -6,7 +6,7 @@ export const createAdvertisement: MutationResolvers["createAdvertisement"] =
   async (_parent, args, _context) => {
     // Creates new Ad.
     const createdAd = await Advertisement.create({
```
Comment (xoldyckk) [inline src/resolvers/Mutation/createAdvertisement.ts:8]: `const createdAdvertisement = await ....`

try to be explicit and precise in naming, there's no harm in writing long self-descriptive names

File: src/typeDefs/inputs.ts [hunk 5]
```diff
@@ -408,4 +408,13 @@ export const inputs = gql`
     imageUrl: String
     videoUrl: String
   }
+
+  input CreateAdvertisementInput {
+    endDate: Date!
+    mediaUrl: URL!
+    name: String!
+    orgId: ID!
```
Comment (xoldyckk) [inline src/typeDefs/inputs.ts:416]: rename to `organizationId: ID!`

File: src/typeDefs/mutations.ts [hunk 6]
```diff
@@ -88,14 +88,7 @@ export const mutations = gql`
...
 
-    createAdvertisement(
-      orgId: ID!
-      name: String!
-      link: String!
-      type: String!
-      startDate: Date!
-      endDate: Date!
-    ): Advertisement!
+    createAdvertisement(data: CreateAdvertisementInput!): Advertisement!
```
Comment (xoldyckk) [inline src/typeDefs/mutations.ts:91]: rename `data` to `input`, the point is to have uniform structure for mutations

here `<mutation_name>` is a placeholder
```graphql
input <mutation_name>Input {
  # any arbitrary object structure required by the mutation, can include scalars, enums and input types 
}

type <mutation_name>Payload {
  # any arbitrary object structure you want to return as the result of this mutation
  # usually would include a single field that the mutation works on, but can be extended to more than
  # one fields
}

type Mutation {
  <mutation_name>(input: <mutation_name(in PascalCase)>Input!): <mutation_name(in PascalCase)>Payload
}
```

here's a real world example, `<mutation_name>` is replaced with `createUser`, `<mutation_name(in PascalCase)>` is replaced with `CreateUser`
```graphql
type User {
  _id: ID!
  age: Int!
  name: String!
}

input CreateUserInput {
  age: Int!
  name: String!
}

type CreateUserPayload {
  accessToken: String!
  refreshToken: String!
  user: User!
}

type Mutation {
  createUser(input: CreateUserInput!): CreateUserPayload
}
```

you see how this would lead to a coherent graphql schema structure for mutations? we can always be sure that each mutation would:-

1. have at least zero and at most one input object defined for it
2. have either zero arguments or exactly one `nullable` argument named `input` with name starting as `mutation(in PascalCase)` and ending with `Input` appended to it
3. return exactly one object type with name starting as `mutation_name(in PascalCase)` and ending with `Payload` appended to it

if you go lax on these standards and conventions, you create a spaghetti hard to understand graphql schema

File: src/typeDefs/types.ts [hunk 7]
```diff
@@ -86,7 +86,7 @@ export const types = gql`
     _id: ID
     name: String!
     orgId: ID
```
Comment (xoldyckk) [inline src/typeDefs/types.ts:88]: rename to `organizationId`, also why is this nullable here?

File: src/models/Advertisement.ts [hunk 8]
```diff
@@ -54,7 +54,7 @@ const advertisementSchema = new Schema({
   orgId: {
     type: String,
```
Comment (xoldyckk) [inline src/models/Advertisement.ts:55]: rename to `organizationId`, why is this nullable in mongoose schema?

File: src/models/Advertisement.ts [hunk 9]
```diff
@@ -11,7 +11,7 @@ export interface InterfaceAdvertisement {
   _id: Types.ObjectId;
   orgId: string;
```
Comment (xoldyckk) [inline src/models/Advertisement.ts:12]: rename to `organizationId`

File: src/typeDefs/mutations.ts (hunk 6 above)
Comment (disha1202) [reply src/typeDefs/mutations.ts]: Thanks for the detailed explanation.

**Comment by disha1202** [comment]
> > > * The aim is to upload an image not use a URL.
> > > * Why is the URL methodology being used?
> > 
> > 
> > Oh Okay, I might have misunderstood earlier.
> > Yes this is suitable for uploading the image/video in advertisement instead of URL.
> > And by URL methodology you mean the type [URL](https://github.com/PalisadoesFoundation/talawa-api/pull/1585/files#diff-efc1675187620595b0844197a25c35eac9b6df752f9182e5cffddee6f27a8489R8)?
> 
> My concern is that it seems that you are using a variable named mediaURL to reference a file upload. Is going to be confusing to future developers.
> 
> Also, is the file stored using the standard methodology for all other uploads? For example, post images.

@palisadoes 

I apologise for the misunderstanding above. 
The mediaUrl variable stores the URL of the media where it is uploaded and not the base 64 string that we are passing in the argument. 

The media is uploaded using [uploadEncodedImage](https://github.com/PalisadoesFoundation/talawa-api/blob/61d4122b769c91d70b59bda26477b35ef773eb81/src/utilities/encodedImageStorage/uploadEncodedImage.ts) and [uploadEncodedVideo](https://github.com/PalisadoesFoundation/talawa-api/blob/61d4122b769c91d70b59bda26477b35ef773eb81/src/utilities/encodedVideoStorage/uploadEncodedVideo.ts)

I have updated the code to use the following for uploading media.

File: schema.graphql [hunk 10]
```diff
@@ -482,7 +497,7 @@ type Mutation {
   cancelMembershipRequest(membershipRequestId: ID!): MembershipRequest!
   checkIn(data: CheckInInput!): CheckIn!
   createAdmin(data: UserAndOrganizationInput!): User!
-  createAdvertisement(endDate: Date!, link: String!, name: String!, orgId: ID!, startDate: Date!, type: String!): Advertisement!
+  createAdvertisement(input: CreateAdvertisementInput!): CreateAdvertisementPayload
```
Comment (EshaanAgg) [inline schema.graphql:500]: No need to return `CreateAdvertismentPayload`. We can return `Advertisement` directly.

File: schema.graphql [hunk 11]
```diff
@@ -96,6 +96,21 @@ type ConnectionPageInfo {
...
+  organizationId: ID!
+  startDate: Date!
+  type: String!
+}
+
+type CreateAdvertisementPayload {
+  _id: ID
+  advertisement: Advertisement!
+}
+
```
Comment (EshaanAgg) [inline schema.graphql:113]: There is no need to create a separate type for it as `Advertisement` with an `_id` field as Mongoose supplies it.

**Review by EshaanAgg — CHANGES_REQUESTED** [review]

File: schema.graphql [hunk 12]
```diff
@@ -96,6 +96,21 @@ type ConnectionPageInfo {
   startCursor: String
 }
 
+
+input CreateAdvertisementInput {
+  endDate: Date!
+  file: String!
```
Comment (xoldyckk) [inline schema.graphql:102]: maybe this is better named `mediaFile`

File: schema.graphql (hunk 10 above)
Comment (xoldyckk) [reply schema.graphql:500]: @EshaanAgg mutations can return more than one type of object, it's not always the case that they mutate a single object and return the same thing, sometimes they can mutate many different types of objects and those different types of mutated objects need to be returned to the client

though not a strict requirement I think they help achieve consistency in the schema's design

File: schema.graphql [hunk 13]
```diff
@@ -96,6 +96,21 @@ type ConnectionPageInfo {
...
+  name: String!
+  organizationId: ID!
+  startDate: Date!
+  type: String!
+}
+
+type CreateAdvertisementPayload {
+  _id: ID
+  advertisement: Advertisement!
```
Comment (xoldyckk) [inline schema.graphql:111]: should just be 
```graphql
type CreateAdverrtisementPayload {
  advertisement: Advertisement
}

File: schema.graphql [hunk 14]
```diff
@@ -27,9 +27,9 @@ input AddressInput {
 type Advertisement {
   _id: ID
   endDate: Date!
-  link: String!
+  mediaUrl: URL!
   name: String!
-  orgId: ID
+  organizationId: ID!
```
Comment (xoldyckk) [inline schema.graphql:32]: this should be `organization: Organization`, graphql schema represents relations between objects, the concept of foreign keys(organizationId) is hidden away in the resolvers(query planning) layer 

accordingly create `organization.ts` resolver in `resolvers/Advertisement/` directory to resolve the organization for an advertisement using `parent.organizationId` within that resolver

**Comment by SiddheshKukade** [comment]
please fix the merge conflicts

File: schema.graphql (hunk 14 above)
Comment (disha1202) [reply schema.graphql:34]: @xoldyckk all the advertisement related resolvers are placed inside the `resolvers/Mutation/` directory. 

Are you suggesting that we should create a `resolvers/Advertisement/` directory and place `createAdvertisement.ts`, `removeAdvertisement.ts`, `updateAdvertisement.ts` and `deleteAdvertisementById.ts` resolvers inside it and also create an `organization.ts` resolver in resolvers/Advertisement/ directory to resolve the organization.

Comment (xoldyckk) [reply schema.graphql:34]: @disha1202 read this https://thomasstep.com/blog/writing-graphql-resolvers

**Comment by palisadoes** [comment]
Please fix the failing tests

**Comment by palisadoes** [comment]
Please fix the failing tests

**Comment by disha1202** [comment]
Hi @palisadoes 

I have fixed the failing tests.

**Comment by disha1202** [comment]
> Can you get the code coverage tests to pass?

I have improved the code coverage.

**Comment by palisadoes** [comment]
This is ready for review

@sumitra19jha
@SiddheshKukade
@xoldyckk
@EshaanAgg

File: src/models/Advertisement.ts [hunk 15]
```diff
@@ -1,15 +1,15 @@
...
 import { Schema, model, models } from "mongoose";
 import type { InterfaceUser } from "./User";
 /**
  * This is an interface that represents a database(MongoDB) document for Advertisement.
  */
 export interface InterfaceAdvertisement {
-  _id: Types.ObjectId;
-  orgId: string;
+  _id: string;
+  organization: any;
```
Comment (xoldyckk) [inline src/models/Advertisement.ts:9]: `organizationId: string`

File: src/models/Advertisement.ts [hunk 16]
```diff
@@ -67,16 +66,17 @@ const advertisementSchema = new Schema(
       type: String,
       required: true,
     },
-    orgId: {
-      type: String,
+    organization: {
```
Comment (xoldyckk) [inline src/models/Advertisement.ts:69]: `organizationId`

File: src/resolvers/Advertisement/organization.ts [hunk 17]
```diff
@@ -0,0 +1,10 @@
+import type { AdvertisementResolvers } from "../../types/generatedGraphQLTypes";
+import { Organization } from "../../models";
+
+export const organization: AdvertisementResolvers["organization"] = async (
+  parent
+) => {
+  return Organization.findOne({
+    _id: parent.organization._id,
```
Comment (xoldyckk) [inline src/resolvers/Advertisement/organization.ts:8]: `parent.organizationId`

File: src/resolvers/Mutation/removeAdvertisement.ts [hunk 18]
```diff
@@ -21,7 +21,7 @@ export const removeAdvertisement: MutationResolvers["removeAdvertisement"] =
```
Comment (xoldyckk) [inline src/resolvers/Mutation/removeAdvertisement.ts:21]: `_id: args.id ? args.id : ""` -> `_id: args.id`

rename `currentAd` to `advertisement` or `existingAdvertisement`

also why is there no check here for who is allowed to delete the advertisement?

File: src/typeDefs/types.ts [hunk 19]
```diff
@@ -135,6 +135,11 @@ export const types = gql`
     advertisement: Advertisement
   }
 
+  type CreateAdvertisementPayload {
+    _id: ID!
```
Comment (xoldyckk) [inline src/typeDefs/types.ts:139]: remove `_id: ID!`, what is this added for?

File: tests/helpers/advertisement.ts [hunk 20]
```diff
@@ -3,35 +3,39 @@ import type { InterfaceUser } from "../../src/models";
 
 import { nanoid } from "nanoid";
 import type { Document } from "mongoose";
+import { createTestUserAndOrganization } from "./userAndOrg";
+import { PopulatedDoc } from "mongoose";
 
 export type TestAdvertisementType = {
   _id: string;
+  organization: any;
```
Comment (xoldyckk) [inline tests/helpers/advertisement.ts:11]: `organizationId: string`

File: src/resolvers/Mutation/createAdvertisement.ts [hunk 21]
```diff
@@ -1,14 +1,50 @@
...
+    }
+
     // Creates new Ad.
-    const createdAd = await Advertisement.create({
-      ...args,
+    const createdAdvertisement = await Advertisement.create({
+      ...args.input,
+      mediaUrl: uploadMediaFile,
       creatorId: _context.userId,
+      organization: args.input.organizationId,
```
Comment (xoldyckk) [inline src/resolvers/Mutation/createAdvertisement.ts:43]: `organizationId`

also you have to check for existence of an organization with `_id === args.input.organizationId`

mongodb doesn't have referential integrity, so it won't check if there's actually an organization with that id while creating the advertisement

File: src/resolvers/Mutation/createAdvertisement.ts [hunk 22]
```diff
@@ -1,14 +1,50 @@
...
+      );
+    }
+
+    let uploadMediaFile = null;
+
+    if (args.input.mediaFile) {
+      const dataUrlPrefix = "data:";
+      if (args.input.mediaFile.startsWith(dataUrlPrefix + "image/")) {
+        uploadMediaFile = await uploadEncodedImage(args.input.mediaFile, null);
+        console.log(uploadMediaFile, "Media File");
```
Comment (xoldyckk) [inline src/resolvers/Mutation/createAdvertisement.ts:30]: remove the console.log

File: src/resolvers/Mutation/updateAdvertisement.ts [hunk 23]
```diff
@@ -82,12 +84,29 @@ export const updateAdvertisement: MutationResolvers["updateAdvertisement"] =
       );
     }
 
+    let uploadMediaFile = null;
+
+    if (args.input.mediaFile) {
+      const dataUrlPrefix = "data:";
+      if (args.input.mediaFile.startsWith(dataUrlPrefix + "image/")) {
+        uploadMediaFile = await uploadEncodedImage(args.input.mediaFile, null);
+        console.log(uploadMediaFile, "Media File");
```
Comment (xoldyckk) [inline src/resolvers/Mutation/updateAdvertisement.ts:93]: remove the console.log

File: src/resolvers/Mutation/createAdvertisement.ts [hunk 24]
```diff
@@ -1,14 +1,50 @@
...
-// @ts-ignore
+import { Advertisement, User } from "../../models";
+import { errors, requestContext } from "../../libraries";
+import { USER_NOT_FOUND_ERROR } from "../../constants";
+import { uploadEncodedImage } from "../../utilities/encodedImageStorage/uploadEncodedImage";
+import { uploadEncodedVideo } from "../../utilities/encodedVideoStorage/uploadEncodedVideo";
+
 export const createAdvertisement: MutationResolvers["createAdvertisement"] =
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars
   async (_parent, args, _context) => {
```
Comment (xoldyckk) [inline src/resolvers/Mutation/createAdvertisement.ts:9]: `_context` -> `context`

underscore is used for unused variables

File: src/resolvers/Mutation/updateAdvertisement.ts [hunk 25]
```diff
@@ -105,8 +124,8 @@ export const updateAdvertisement: MutationResolvers["updateAdvertisement"] =
     const updatedAdvertisementPayload = {
       _id: updatedAdvertisement._id.toString(), // Ensure _id is converted to String as per GraphQL schema
       name: updatedAdvertisement.name,
-      orgId: updatedAdvertisement.orgId.toString(),
-      link: updatedAdvertisement.link,
+      organization: updatedAdvertisement.organization,
```
Comment (xoldyckk) [inline src/resolvers/Mutation/updateAdvertisement.ts:127]: `organizationId`

File: src/resolvers/Query/getAdvertisements.ts [hunk 26]
```diff
@@ -5,7 +5,18 @@ import { Advertisement } from "../../models";
  * This function returns list of Advertisement from the database.
  * @returns An object that contains a list of Ads.
  */
-export const getAdvertisements: QueryResolvers["getAdvertisements"] =
-  async () => {
-    return await Advertisement.find().lean();
-  };
+export const getAdvertisements: QueryResolvers["getAdvertisements"] = async (
```
Comment (xoldyckk) [inline src/resolvers/Query/getAdvertisements.ts:8]: rename this to just `advertisements` in resolvers and graphql schema

**Review by SiddheshKukade — CHANGES_REQUESTED** [review]
I can see that there are multiple renames for schemas. So is it working fine with the frontend ? Or Should I create an issue for that ?

**Comment by disha1202** [comment]
> I can see that there are multiple renames for schemas. So is it working fine with the frontend ? Or Should I create an issue for that ?

All the changes made in the schema are addressed in this [PR](https://github.com/PalisadoesFoundation/talawa-admin/pull/1529)

File: src/resolvers/Mutation/removeAdvertisement.ts (hunk 18 above)
Comment (disha1202) [reply src/resolvers/Mutation/removeAdvertisement.ts:21]: Ads management is done by ADMIN and SUPER ADMIN and USER can not access the admin portal.
So the check for who is allowed to delete an advertisement shouldn't be required.

What do you think?

Also the `removeAdvertisement` Mutation is not being used anywhere instead `src/resolvers/Mutation/deleteAdvertisementById.ts` is used. 

Can we update the `deleteAdvertisementById` mutation and remove the `removeAdvertisement` Mutation?

**Comment by palisadoes** [comment]
Please fix the failing tests

File: src/resolvers/Mutation/removeAdvertisement.ts (hunk 18 above)
Comment (xoldyckk) [reply src/resolvers/Mutation/removeAdvertisement.ts:21]: it's not about the admin portal, it's about general security, you're assuming that no user would ever make a `removeAdvertisement` graphql mutation request through a portal other than admin portal, that's a naive assumption

yeah sure, get rid of one of them, also if `deleteAdvertisementById` is kept, rename it to `deleteAdvertisement`

**Comment by disha1202** [comment]
@xoldyckk @palisadoes 

Currently we are fetching all the advertisements using the `advertisements` query resolver and filtering the advertisements by `organizationId`. 

Should we instead have a `getAdvertisementsByOrgId` query resolver which returns the advertisements of a particular organization?

**Comment by xoldyckk** [comment]
@disha1202 
```graphql
type AdvertisementEdge {
  cursor: String
  node: Advertisement
}

type AdvertisementsConnection {
  edges: [AdvertisementEdge]
  pageInfo: ConnectionPageInfo
  totalCount: Int
}

type Organization {
  advertisements(after: String, before: String, first: Int, last: Int): AdvertisementsConnection
}
```

**Comment by palisadoes** [comment]
Please fix the conflicting files

**Comment by SiddheshKukade** [comment]
@disha1202 fix conflicts

**Comment by palisadoes** [comment]
*NOTE* Read very carefully

1. We just merged this PR which upgraded the prettier package. 
    1. https://github.com/PalisadoesFoundation/talawa-api/pull/1840
1. It reformatted over 400 files. This will put your PR at risk of extensive merge conflicts.    
1. I suggest that before you merge your code to the latest upstream that you:
    1. upgrade your `prettier` in your local branch to the same version
    1. run `prettier`
    
This will help to reduce the number of future merge conflicts for your PR.

**Comment by SiddheshKukade** [comment]
@disha1202 please take a pull from origin develop

**Comment by palisadoes** [comment]
@disha1202 

This has been open for two months with numerous commits but no approvals.

Is there some difficulty?

**Comment by disha1202** [comment]
> @disha1202
> 
> This has been open for two months with numerous commits but no approvals.
> 
> Is there some difficulty?

@palisadoes Initially the scope of the PR was to add support to upload media files while creating an advertisement. 
But we've also refactored and optimised the overall functionality which is why it has been going on since 2 months. 

I apologise for the delay, I'll try to wrap this up as soon as possible.

**Comment by disha1202** [comment]
> @disha1202 Is this ready? There is a failed Testing Application check.

Hi @Cioppolo14 

I have fixed the failing tests.

File: codegen.ts [hunk 27]
```diff
@@ -30,6 +30,14 @@ const config: CodegenConfig = {
           ActionItemCategory:
             "../models/ActionItemCategory#InterfaceActionItemCategory",
 
+          Advertisement: "../models/Advertisement#InterfaceAdvertisement",
+
+          CreateAdvertisementPayload:
+            "../models/Advertisement#InterfaceAdvertisement",
+
+          UpdateAdvertisementPayload:
+            "../models/Advertisement#InterfaceAdvertisement",
```
Comment (xoldyckk) [inline codegen.ts:39]: Except the `Advertisement` mapping remove the other two, this configuration is only for mapping database model types to corresponding graphql types

File: src/typeDefs/inputs.ts [hunk 28]
```diff
@@ -459,4 +459,13 @@ export const inputs = gql`
     imageUrl: String
     videoUrl: String
   }
+
+  input CreateAdvertisementInput {
+    endDate: Date!
+    name: String!
+    organizationId: ID!
+    startDate: Date!
+    type: String!
```
Comment (xoldyckk) [inline src/typeDefs/inputs.ts:468]:
```graphql
enum AdvertisementType {
  POP_UP
  MENU
  BANNER
}
```
declare this and use this as the advertisement's `type` field everywhere in the schema

File: src/typeDefs/queries.ts [hunk 29]
```diff
@@ -67,7 +67,7 @@ export const queries = gql`
     getlanguage(lang_code: String!): [Translation]
 
     getPlugins: [Plugin]
-    getAdvertisements: [Advertisement]
+    advertisements: [Advertisement]
```
Comment (xoldyckk) [inline src/typeDefs/queries.ts:70]: you are already exposing advertisements on Organization, is this field necessary? One use case would be accessing all advertisements for all organizations. If that is the case make this field `AdvertisementsConnection` too.

File: src/typeDefs/types.ts [hunk 30]
```diff
@@ -151,10 +151,34 @@ export const types = gql`
...
+    node: Advertisement
+  }
+
+  type AdvertisementsConnection {
+    edges: [AdvertisementEdge]
+    pageInfo: ConnectionPageInfo
+    totalCount: Int
+  }
+
+  extend type Organization {
```
Comment (xoldyckk) [inline src/typeDefs/types.ts:165]: You have exposed `Organization.advertisements` two times in the same schema, delete this one.

File: src/utilities/parseRelayConnectionArguments.ts [hunk 31]
```diff
@@ -0,0 +1,91 @@
+import { GraphQLError } from "graphql";
```
Comment (xoldyckk) [inline src/utilities/parseRelayConnectionArguments.ts:1]: delete this file, and use the function exported from `src/utilities/graphQLConnection/parseGraphQLConnectionArguments.ts`

File: src/typeDefs/mutations.ts [hunk 32]
```diff
@@ -117,21 +117,16 @@ export const mutations = gql`
...
+    ): CreateAdvertisementPayload @auth
 
     createPost(data: PostInput!, file: String): Post @auth
 
     createUserTag(input: CreateUserTagInput!): UserTag @auth
 
     createSampleOrganization: Boolean! @auth
 
-    deleteAdvertisementById(id: ID!): DeletePayload!
+    deleteAdvertisement(id: ID!): DeletePayload!
```
Comment (xoldyckk) [inline src/typeDefs/mutations.ts:129]:
```graphql
type DeleteAdvertisementPayload {
  advertisement: Advertisement
}

deleteAdvertisement(id: ID!): DeleteAdvertisementPayload
```

Comment (SiddheshKukade) [reply src/typeDefs/mutations.ts:135]: @xoldyckk actually the code is just returning the success message. So the current payload is fine.
```    // Deletes the ad.
    const deleteAdvertisement = await Advertisement.deleteOne({
      _id: args.id,
    });
    // Returns deleted ad.
    return { success: deleteAdvertisement.deletedCount ? true : false };
```


https://github.com/disha1202/talawa-api/blob/43f9ad971a42d3333bf086e568adbc4260a77d6e/src/resolvers/Mutation/deleteAdvertisement.ts#L45-L50

File: schema.graphql [hunk 33]
```diff
@@ -68,20 +68,31 @@ type Advertisement {
...
   BANNER
   MENU
   POPUP
 }
 
+type AdvertisementsConnection {
+  edges: [AdvertisementEdge]
+  pageInfo: ConnectionPageInfo
+  totalCount: Int
+}
```
Comment (SiddheshKukade) [inline schema.graphql:94]: Can you please also update the PR description explaining what AdvertisementEdge and Connection means in layman terms.
Thanks
@disha1202

**Comment by palisadoes** [comment]
Why is the JSDocs test being skipped?

**Review by SiddheshKukade — DISMISSED** [review]
@disha1202 please do the minor fixes suggested by @xoldyckk Other than that, the code looks fine. 
Please also add the details about the parts that I've mentioned. in the PR description

**Comment by SiddheshKukade** [comment]
> Why is the JSDocs test being skipped?

@palisadoes I've updated the branch to latest. Should work now.

**Comment by palisadoes** [comment]
@disha1202 

We need the JSDocs tests to pass. Please update.

File: src/typeDefs/mutations.ts (hunk 32 above)
Comment (disha1202) [reply src/typeDefs/mutations.ts]: I updated the code to return the deleted advertisement.

**Comment by xoldyckk** [comment]
@palisadoes why was this merged without approval?

**Comment by palisadoes** [comment]
1. It was approved a few hours ago by @SiddheshKukade 
2. It was dismissed when the tests were updated with subsequent commits


image

File: src/resolvers/Mutation/createAdvertisement.ts [hunk 34]
```diff
@@ -1,14 +1,68 @@
...
+
+    // Checks whether organization with _id == args.data.organizationId exists.
+    if (!organization) {
+      throw new errors.NotFoundError(
+        requestContext.translate(ORGANIZATION_NOT_FOUND_ERROR.MESSAGE),
+        ORGANIZATION_NOT_FOUND_ERROR.CODE,
+        ORGANIZATION_NOT_FOUND_ERROR.PARAM,
+      );
+    }
+
```
Comment (xoldyckk) [inline src/resolvers/Mutation/createAdvertisement.ts:39]: Any normal user can create an advertisement, is this expected?

File: src/resolvers/Mutation/deleteAdvertisement.ts [hunk 35]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    if (currentUser.userType === "USER") {
+      throw new errors.UnauthorizedError(
+        requestContext.translate(USER_NOT_AUTHORIZED_ERROR.MESSAGE),
+        USER_NOT_AUTHORIZED_ERROR.CODE,
+        USER_NOT_AUTHORIZED_ERROR.PARAM,
+      );
+    }
+
```
Comment (xoldyckk) [inline src/resolvers/Mutation/deleteAdvertisement.ts:31]: Here, admin of any organization can delete any advertisement, is this expected?

File: src/resolvers/Mutation/updateAdvertisement.ts [hunk 36]
```diff
@@ -82,12 +83,28 @@ export const updateAdvertisement: MutationResolvers["updateAdvertisement"] =
       );
     }
```
Comment (xoldyckk) [inline src/resolvers/Mutation/updateAdvertisement.ts:85]: Here superadmins, and admins of any organization can update any advertisement, is this expected?

File: src/resolvers/Query/advertisementsConnection.ts [hunk 37]
```diff
@@ -0,0 +1,22 @@
+import type { QueryResolvers } from "../../types/generatedGraphQLTypes";
```
Comment (xoldyckk) [inline src/resolvers/Query/advertisementsConnection.ts:1]: @disha1202  Just the name of the file was changed, the internal logic wasn't changed to be a connection.

**Comment by disha1202** [comment]
Should I create a new PR to address the following issues?

**Comment by palisadoes** [comment]
Yes.
